src/dns/coredns.rs

```diff
@@ -509,3 +521,103 @@ fn reply_ip<'a>(
     }
     Some(req)
 }
+
+#[cfg(test)]
+mod test {
+    use std::{
+        fmt::Display,
+        future::ready,
+        io,
+        sync::OnceLock,
+        task::{Context, Poll},
+    };
+
+    use clap::error::ErrorKind;
+    use hickory_proto::{op::Query, xfer::DnsRequestSender};
+    use tokio::time::error::Elapsed;
+
+    use super::*;
+
+    struct TestUdp;
+
+    #[derive(Debug)]
+    struct BrokenPipeError;
+
+    impl Display for BrokenPipeError {
+        fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+            write!(f, "BrokenPipeError")
+        }
+    }
+
+    impl std::error::Error for BrokenPipeError {}
+
+    impl DnsUdpSocket for TestUdp {
+        type Time = hickory_proto::TokioTime;
+
+        fn poll_recv_from(
+            &self,
+            cx: &mut Context<'_>,
+            buf: &mut [u8],
+        ) -> Poll<io::Result<(usize, SocketAddr)>> {
+            Poll::Ready(io::Result::Err(Error::new(
+                std::io::ErrorKind::BrokenPipe,
+                BrokenPipeError,
+            )))
+        }
+
+        fn poll_send_to(
+            &self,
+            cx: &mut Context<'_>,
+            buf: &[u8],
+            target: SocketAddr,
+        ) -> Poll<io::Result<usize>> {
+            unimplemented!()
+        }
+    }
+
+    impl CoreDnsUdp for TestUdp {
+        fn local_addr(&self) -> std::io::Result<SocketAddr> {
+            Ok("127.0.0.1:0".parse().unwrap())
+        }
+    }
+
+    // we need 2 threads or tokio::spawn will block since it never yields
+    #[test_log::test(tokio::test(flavor = "multi_thread", worker_threads = 2))]
```

Comment:
What is the point of test_log? It seem this can just be dropped so we do not need extra dependencies.